src/docfx/build/toc/ServicePageGenerator.cs

```diff
@@ -57,41 +51,61 @@ internal class ServicePageGenerator
                     servicePagePath = FilePath.Generated(new PathString($"./{baseDir}/{directoryName}/{filename}.yml"));
                 }
 
+                if (string.IsNullOrEmpty(node.Href.Value))
+                {
+                    var servicePageFullPath = Path.GetFullPath(Path.Combine(_docsetPath, servicePagePath.Path)) ?? _docsetPath;
+                    var referenceTocFullPath = Path.GetDirectoryName(Path.GetFullPath(Path.Combine(_docsetPath, _joinTOCConfig.ReferenceToc ?? ".")));
+                    var hrefRelativeToReferenceTOC = Path.GetRelativePath(referenceTocFullPath ?? ".", servicePageFullPath);
+                    node.Href = node.Href.With(hrefRelativeToReferenceTOC);
+                }
+
                 var name = node.Name;
                 var fullName = node.Name;
-                var uid = node.Uid;
-                var href = node.Href;
-
+                node.Uid = node.Uid.With(null);
                 var children = new List<ServicePageItem>();
                 foreach (var item in node.Items)
                 {
                     ServicePageItem child;
                     var childName = item.Value.Name.Value;
                     var childHref = item.Value.Href.Value;
                     var childUid = item.Value.Uid.Value;
-
                     var childHrefType = TableOfContentsLoader.GetHrefType(childHref);
-
                     if (!string.IsNullOrEmpty(childHref) && (childHrefType == TocHrefType.RelativeFolder || childHrefType == TocHrefType.TocFile))
                     {
                         childHref = null;
                     }
 
-                    if (!string.IsNullOrEmpty(childHref))
+                    if (item.Value.LandingPageType.Value != null)
                     {
-                        if (!(childHref.StartsWith("~/") || childHref.StartsWith("~\\")))
+                        if (!string.IsNullOrEmpty(childHref))
```

Comment:
Would it be clearer to put `if (!string.IsNullOrEmpty(childHref))` as outer check?
The check of `if (item.Value.LandingPageType.Value != null)` only affects the value of `childHref`, maybe we can put it into `GetHrefRelativeToServicePage`?

Reply:
If we do that, we still cannot put `if (item.Value.LandingPageType.Value != null)` into `GetHrefRelativeToServicePage`. Because when `childHref` isn't empty, we also need to generated href instead of using uid when children are also service pages.